azure-kusto-data/azure/kusto/data/security.py

```diff
@@ -116,57 +79,100 @@ def __init__(self, kcsb):
         aad_authority_uri = os.environ.get("AadAuthorityUri", CLOUD_LOGIN_URL)
         self.authority_uri = aad_authority_uri + authority if aad_authority_uri.endswith("/") else aad_authority_uri + "/" + authority
 
-    def acquire_authorization_header(self):
+    def _build_kusto_authentication_error(self, error: KustoAuthenticationError) -> KustoAuthenticationError:
+        if self.authentication_method is AuthenticationMethod.aad_username_password:
+            kwargs = {"username": self.username, "client_id": self.client_id}
+        elif self.authentication_method is AuthenticationMethod.aad_application_key:
+            kwargs = {"client_id": self.client_id}
+        elif self.authentication_method is AuthenticationMethod.aad_device_login:
+            kwargs = {"client_id": self.client_id}
+        elif self.authentication_method is AuthenticationMethod.aad_application_certificate:
+            kwargs = {"client_id": self.client_id, "thumbprint": self.thumbprint}
+        elif self.authentication_method is AuthenticationMethod.aad_msi:
+            kwargs = self.msi_params
+        else:
+            return error
+        kwargs["resource"] = self.kusto_uri
+        if self.authentication_method is AuthenticationMethod.aad_msi:
+            kwargs["authority"] = AuthenticationMethod.aad_msi.value
+        elif self.auth_context is not None:
+            kwargs["authority"] = self.auth_context.authority.url
+        return KustoAuthenticationError(self.authentication_method.value, error, **kwargs)
+
+    @staticmethod
+    def _get_header(token_type, access_token):
+        return "{0} {1}".format(token_type, access_token)
+
+    def _get_header_from_dict(self, token: dict):
+        return self._get_header(token[TokenResponseFields.TOKEN_TYPE], token[TokenResponseFields.ACCESS_TOKEN])
+
+    def _set_from_stored_token(self, stored_token):
+        if (
+            TokenResponseFields.REFRESH_TOKEN in stored_token
+            and TokenResponseFields._CLIENT_ID in stored_token
+            and TokenResponseFields._AUTHORITY in stored_token
+        ):
+            self.client_id = stored_token[TokenResponseFields._CLIENT_ID]
+            self.username = stored_token[TokenResponseFields.USER_ID]
+            self.authority_uri = stored_token[TokenResponseFields._AUTHORITY]
+
+    @staticmethod
+    def _get_expiration_data_from_token(token):
+        return dateutil.parser.parse(token[TokenResponseFields.EXPIRES_ON])
+
+    @staticmethod
+    def _authentication_method_missing_exception() -> KustoClientError:
+        return KustoClientError("Please choose authentication method from azure.kusto.data.security.AuthenticationMethod")
+
+    @staticmethod
+    def _is_expired(expiration_date):
```

Comment:
don't hide the logic. nobody knows what expired means. this method should be removed

Reply:
I prefer that the logic of expiration will be written once... and not twice

Reply:
if you want to, then you need to promote time to be a parameter, with a default value like:
`is_expired(expiration_date, expiry_time=1 * 60`

I'm sure you're familiar with python's zen (explicit is better than implicit)